crates/kona-providers/src/blob_provider.rs

```diff
@@ -0,0 +1,152 @@
+    ///
+    /// The `WithFallback` setup allows to specify two different
+    /// endpoints for a primary and a fallback blob provider.
+    online: OnlineBlobProviderWithFallback<
+        OnlineBeaconClient,
+        OnlineBeaconClient,
+        SimpleSlotDerivation,
+    >,
+}
+
+/// A blob provider that hold blobs in memory.
+#[derive(Debug)]
+pub struct InnerBlobProvider {
+    /// Maximum number of blobs to keep in memory.
+    capacity: usize,
+    /// Order of key insertion for oldest entry eviction.
+    key_order: VecDeque<B256>,
+    /// Maps block hashes to blobs.
+    blocks_to_blob: HashMap<B256, Vec<Blob>>,
+}
+
+impl InnerBlobProvider {
+    /// Creates a new [InnerBlobProvider].
+    pub fn with_capacity(cap: usize) -> Self {
+        Self {
+            capacity: cap,
+            blocks_to_blob: HashMap::with_capacity(cap),
+            key_order: VecDeque::with_capacity(cap),
+        }
+    }
+
+    /// Inserts multiple blobs into the provider.
+    pub fn insert_blobs(&mut self, block_hash: B256, blobs: Vec<Blob>) {
+        if let Some(existing_blobs) = self.blocks_to_blob.get_mut(&block_hash) {
+            existing_blobs.extend(blobs);
+        } else {
+            if self.blocks_to_blob.len() >= self.capacity {
+                if let Some(oldest) = self.key_order.pop_front() {
+                    self.blocks_to_blob.remove(&oldest);
+                }
+            }
+            self.blocks_to_blob.insert(block_hash, blobs);
+        }
+    }
+}
+
+impl LayeredBlobProvider {
+    /// Creates a new [LayeredBlobProvider] with a local blob store, an online primary beacon
+    /// client and an optional fallback blob archiver for fetching blobs.
+    pub fn new(beacon_client_url: Url, blob_archiver_url: Option<Url>) -> Self {
+        let memory = Arc::new(Mutex::new(InnerBlobProvider::with_capacity(512)));
+
+        let online = OnlineBlobProviderBuilder::new()
+            .with_primary(beacon_client_url.to_string())
+            .with_fallback(blob_archiver_url.map(|url| url.to_string()))
+            .build();
+
+        Self { memory, online }
+    }
+
+    /// Inserts multiple blobs into the in-memory provider.
+    #[inline]
+    pub fn insert_blobs(&mut self, block_hash: B256, blobs: Vec<Blob>) {
+        self.memory.lock().insert_blobs(block_hash, blobs);
+    }
+
+    /// Attempts to fetch blobs using the in-memory blob store.
+    #[inline]
+    async fn memory_blob_load(
+        &mut self,
+        block_ref: &BlockInfo,
+        hashes: &[IndexedBlobHash],
+    ) -> eyre::Result<Vec<Blob>> {
+        let locked = self.memory.lock();
+
+        let blobs_for_block = locked
+            .blocks_to_blob
+            .get(&block_ref.hash)
+            .ok_or_else(|| eyre!("Blob not found for block ref: {:?}", block_ref))?;
+
+        let mut blobs = Vec::new();
+        for _ in hashes {
+            for blob in blobs_for_block {
+                blobs.push(*blob);
+            }
+        }
```

Comment:
not sure I understand this double loop

Reply:
ah I'm not 100% sure either as I didn't write this part originally, but my guess is that we returned a copy of all blobs for each index because we didn't have a way to filter blobs by their hashes here.

Now after looking at it more I think we could keep the `BlobTransactionSidecar`s in the store 
(instead of just blob data) and get the hashes from that. It should work given that we try to insert manually 
in the store only when reaching for the `ExExContext::pool` here: https://github.com/paradigmxyz/reth-exex-examples/pull/12/files#diff-fb2ecd571a8b660d914eac307e6aed9fc4fa610ad0b5521145d0fd6115e963dbR210

Reply:
it would look like this if we had the full `BlobTransactionSidecar`s:

```rust
        let sidecars_for_block = locked
            .blocks_to_blob_sidecars
            .get(&block_ref.hash)
            .ok_or(eyre!("No blob sidecars found for block ref: {:?}", block_ref))?;

        // for each sidecar, get the blob hashes and check if any of them are
        // part of the requested hashes. If they are, add the corresponding blobs
        // to the blobs vector.
        let mut blobs = Vec::with_capacity(hashes.len());
        let requested_hashes = hashes.iter().map(|h| h.hash).collect::<Vec<_>>();
        for sidecar in sidecars_for_block {
            for (hash, blob) in sidecar.versioned_hashes().zip(&sidecar.blobs) {
                if requested_hashes.contains(&hash) {
                    blobs.push(*blob);
                }
            }
        }
```

wdyt @refcell @Rjected?

Reply:
that makes sense to me

Reply:
Yes this is correct - thanks for looking into this!